In [1]:

import tensorflow as tf
print(tf.__version__)


2.12.0


In [2]:
# load dataset
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train = x_train/255.0
x_test = x_test/255.0


11490434/11490434 [==============================] - 2s 0us/step


In [3]:
# build a machine learning model
model = tf.keras.models.Sequential([
    tf.keras.layers.Flatten(input_shape = (28,28)),
    tf.keras.layers.Dense(128,activation = 'relu'),
    tf.keras.layers.Dropout(0.2),
    tf.keras.layers.Dense(10)
])
# Sequential: is useful for stacking layers where each layer has one input tensos and one output tensor
# Flatten, Dense, Dropout layers


In [6]:
# for example, the model returns a vector of logits or log-odds scores, one for each class
import numpy as np
predictions = model(x_train[:1]).numpy()
predictions

array([[-0.22213498,  0.09326604,  0.97866386,  1.0177459 , -0.20374681,
        -0.19017819, -0.52295935,  0.16741845,  0.90927225, -0.18595552]],
      dtype=float32)

In [7]:
# the tf.nn.softmax function converts these logits to probability for each class
tf.nn.softmax(predictions).numpy()

array([[0.05696968, 0.07809454, 0.18929714, 0.19684172, 0.05802693,
        0.05881964, 0.0421694 , 0.08410555, 0.1766069 , 0.05906854]],
      dtype=float32)

In [8]:
# define a loss function for training using
loss_fn = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)
#This loss is equal to the negative log probability of the true class:
# the loss is zero if the model is sure of the correct class
# this untrained model gives probabilities close to random (1/10 for each class)
#so the initial loss should be close `-tf.math.log(1/10)~=2.3`
loss_fn(y_train[:1], predictions).numpy()

2.8332794

In [9]:
# configure and compile the model using Model.complile. Set the optimizer clas
# to adam, the loss to the loss-fn and specify a metric to be evaluated
# for the model by setting th metrics parameter to accuracy

model.compile(optimizer='adam',
              loss= loss_fn,
              metrics=['accuracy'])

In [10]:
# train and evaluate our model

model.fit(x_train, y_train, epochs = 5)

Epoch 1/5
1875/1875 [==============================] - 8s 3ms/step - loss: 0.2986 - accuracy: 0.9143
Epoch 2/5
1875/1875 [==============================] - 5s 2ms/step - loss: 0.1440 - accuracy: 0.9570
Epoch 3/5
1875/1875 [==============================] - 5s 3ms/step - loss: 0.1072 - accuracy: 0.9673
Epoch 4/5
1875/1875 [==============================] - 5s 2ms/step - loss: 0.0877 - accuracy: 0.9727
Epoch 5/5
1875/1875 [==============================] - 5s 3ms/step - loss: 0.0738 - accuracy: 0.9771


In [11]:
# The model.evaluate method checks the model's performance
model.evaluate(x_test, y_test, verbose = 2)

313/313 - 1s - loss: 0.0677 - accuracy: 0.9794 - 997ms/epoch - 3ms/step


[0.06770147383213043, 0.9793999791145325]

This image classifier is now trained to 98% accuracy on this dataset. You have a trained a machine learning model using a prebulit dataset using the Keras API
